plasmapy/plasma/grids.py

```diff
@@ -921,65 +943,43 @@ def _interp_units(self):
 
         return _interp_units
 
-    @property
-    def interpolator(self):
+    def nearest_neighbor_interpolator(
+        self, pos: Union[np.ndarray, u.Quantity], *args, persistent=False
+    ):
         r"""
-        A nearest-neighbor interpolator that returns the nearest grid index
-        to a position.
-        """
-        if self._interpolator is None:
-            if self.is_uniform:
-                self._make_uniform_grid_interpolator()
-            else:
-                self._make_nonuniform_grid_interpolator()
+        Interpolate values on the grid using a nearest-neighbor scheme with
+        no higher-order weighting.
 
-        return self._interpolator
+        Parameters
+        ----------
+        pos : `~numpy.ndarray` or `~astropy.units.Quantity` array, shape (n,3)
+            An array of positions in space, where the second dimension
+            corresponds to the three dimensions of the grid. If an
+            `~numpy.ndarray` is provided, units will be assumed to match
+            those of the grid.
 
-    def _make_uniform_grid_interpolator(self):
-        r"""
-        Initializes a nearest-neighbor interpolator that returns the nearest
-        grid indices for a given position (given in SI units).
+        *args : `str`
+            Strings that correspond to DataArrays in the dataset
 
-        This function works on a uniformly spaced grid.
+        persistent : `bool`
+            If `True`, the interpolator will assume the grid and its
+            contents have not changed since the last interpolation. This
+            substantially speeds up the interpolation when many
+            interpolations are performed on the same grid in a loop.
+            ``persistent`` overrides to `False` if the arguments list
+            has changed since the last call.
         """
-        # Create a grid of indices for use in interpolation
-        n0, n1, n2 = self.shape
-        indgrid = np.indices([n0, n1, n2])
-        indgrid = np.moveaxis(indgrid, 0, -1)
-
-        # Create an input array of grid positions
-        pts = (
-            self.ax0.si.value,
-            self.ax1.si.value,
-            self.ax2.si.value,
-        )
-
-        self._interpolator = interp.RegularGridInterpolator(
-            pts, indgrid, method="nearest", bounds_error=False, fill_value=np.nan
+        raise NotImplementedError(
+            "Nearest-neighbor interpolator is not yet "
+            "implemented for this grid type."
```

Comment:
Could/should this `nearest_neighbor_interpolator`, then, be an [abstractmethod](https://docs.python.org/3/library/abc.html#abc.abstractmethod)? In other words - what minimal set of promises do we make about each actual implementation of the AbstractGrid?

Just food for thought for now, I'm okay with this as it is :)

Reply:
Yes, I think probably it should (not the volume_averaged_interpolator, though: those aren't defined for all grids)

Reply:
Ok, so I made `nearest_neighbor_interpolator` an abstract method. I also did the same with `vector_intersects` because the current implementation is inherently Cartesian.

As a result, one can no longer instantiate instances of AbstractGrid. So, in the tests I changed all the AbstractGrid tests to us CartesianGrid, which is a subclass.

Reply:
Seems to fit the Abstract in AbstractGrid, anyway!